Спринт №16

Keithcutt opened the pull request:
Рефакторинг приложения на Clean Architecture.

Keithcutt committed:
Домашнее задание №16

Работа с Retrofit перенесена в data-слой и была протестирована.

Keithcutt committed:
Работа с SharedPreferences перенесена в data-слой и была протестирована.

Keithcutt committed:
Код из SearchActivity частично разнесен между ViewModel и Activity:
- Создана ViewModel для экрана поиска.
- Созданы состояния экрана поиска.

Keithcutt committed:
Логика из SearchActivity вынесена во ViewModel.

Keithcutt committed:
Логика из PlayerActivity вынесена во ViewModel и протестирована.

Также было убрано отображение истории поиска при пустом тексте запроса (вида "     ").

Keithcutt committed:
Работа по смене и сохранению темы приложения была вынесена в data-слой. Теперь взаимодействие с этими функциями происходит через интерактор. Работа с domain-слоем происходит во вьюмодели.

А также выполнен небольшой рефакторинг методов onCreate в Activity экрана плеера и поиска.

Keithcutt committed:
Оставшиеся опции экрана настроек были переписаны с использованием чистой архитектуры аналогично настройкам темы.

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/settings/data/repository/SettingsRepository.kt:
```diff
@@ -0,0 +1,8 @@
+
+import com.example.playlistmaker.settings.domain.model.ThemeSettings
+
+interface SettingsRepository {
```
Вот этот интерфейс должен быть в domain слое объявлен, чтобы не нарушать правило зависимостей (domain не должен зависеть от data)

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/settings/domain/impl/SettingsInteractorImpl.kt:
```diff
@@ -0,0 +1,17 @@
+package com.example.playlistmaker.settings.domain.impl
+
+import com.example.playlistmaker.settings.data.repository.SettingsRepository
```
Вот тут нарушается правило зависимостей - класс из domain слоя зависит от интерфейса из data, это пофиксится, когда перенесёшь репозиторий

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/sharing/domain/impl/SharingInteractorImpl.kt:
```diff
@@ -0,0 +1,22 @@
+package com.example.playlistmaker.sharing.domain.impl
+
+import com.example.playlistmaker.sharing.data.ExternalNavigator
```
Здесь тоже нарушается правило зависимостей, интерфейс нужно положить в domain слой

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/search/data/dto/TrackSearchResponse.kt:
```diff
@@ -0,0 +1,4 @@
+package com.example.playlistmaker.search.data.dto
+
+data class TrackSearchResponse(val resultCount: Int,
```
Тут можно не делать data класс, фичи дата класса нигде не используются, поэтому просто генерируется лишний код, можно просто `class`

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/search/presentation/state/SearchScreenState.kt:
```diff
@@ -0,0 +1,12 @@
+sealed interface SearchScreenState{
+    data object Loading: SearchScreenState
+    class SearchQueryResults(val foundTracks: MutableList<Track>): SearchScreenState
+    class SearchHistory(val searchedTracks: MutableList<Track>): SearchScreenState
```
На ui лучше не отдавать `MutableList`, всегда лучше отдавать копии списков в виде интерфейса `List`

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/search/ui/SearchViewHolder.kt:
```diff
@@ -8,28 +8,30 @@ import android.widget.TextView
-    fun bind(model: Track) {
+    fun bind(currentTrack: Track) {
+
+        val track = BriefTrackInfoMapper.map(currentTrack)
```
А зачем здесь маппить в другую модель, которая содержит меньше данных, чем оригинальная домейн модель? Даже если и маппить, то маппинг нужно выполнять во вьюмодели, а не при отрисовке ui

ArturNurtdinov reviewed (commented):
Бэкграунд экрана поиска не соответствует макетам, плюс бэкраунд экрана поиска отличается от цвета тулбара:
![image](https://github.com/user-attachments/assets/a5ab5ee2-2b8b-4e2d-8d89-546d38f34fc2)
По макетам цвет в тёмной теме должен быть #1A1B22 (в белой теме - белый, тоже нужно поправить)

Keithcutt committed:
Выполнена работа над ошибками по результатам ревью.

Также был исправлен баг с отображением интерфейса плеера при повороте устройства в режиме воспроизведения.
